### ISSUE TITLE
Why are the interfaces for `Insert` and `Merge` different for multiple inserts/merges ?

### ISSUE THREAD

**Issue by juanmirocks** [jOOQ/jOOQ#10634]
### Your question:

I use scala & jooq (for many years already, and I'm super happy with it -- thank you for this awesome software!!!).

**TL;DR**: the jooq interfaces for `insertInto` (insert) and `mergeInto` (merge) are different for multiple inserts. Why? Are there workarounds or better patterns?

Long explanation:

Sometimes I use this pattern when I want to do **multiple inserts**:

```scala
// Prepare `insertInto` statement
var query =
  (sql //also often called `create` in jooq documentation
    insertInto(MY_TABLE, ...))

// Then add the values; we could do this with a fold, but not relevant for this question
myValues.foreach { ... =>
  query = query.values(...)
}

// Finally execute
query.execute()
```

That works great for inserts. However, **the same pattern does not work for `mergeInto`**:

```scala
// Prepare `mergeInto` statement
var query =
  (sql //also often called `create` in jooq documentation
    mergeInto(MY_TABLE, ...))

// Then add the values; we could do this with a fold, but not relevant for this question
myValues.foreach { ... =>
  query = query.values(...) //**COMPILATION ERROR**
}

...
```

The line `query = query.values(...)` does not compile. The error is something like:

```shell
type mismatch;
[error]  found   : org.jooq.Merge[...]
[error]  required: org.jooq.MergeValuesStep4[...]
```

I understand the compilation error, and understand that the interfaces of [Merge](https://www.jooq.org/javadoc/3.11.x/org/jooq/Merge.html) and [MergeValuesStepN](https://www.jooq.org/javadoc/3.11.x/org/jooq/MergeValuesStep4.html) are different.

My question is: why the same pattern works for `insertInto`, whereas it does not work for `mergeInto`. Is there a particular design for this, or this problem just didn't itch anybody yet?

On top of my question: am I perhaps missing something and there is a better pattern to achieve the same (namely multiple inserts/merges)? 

### Versions:

- jOOQ: 3.11.11 and 3.13.4
- DB: h2
- _rest of versions are not relevant for this question_

**Comment by lukaseder** [comment]
Thank you very much for your question and for your nice words. I'm happy to hear you've found jOOQ so useful in the past.

> **TL;DR**: the jooq interfaces for `insertInto` (insert) and `mergeInto` (merge) are different for multiple inserts. Why? Are there workarounds or better patterns?

jOOQ mimicks the SQL language as much as possible. In native SQL, the statements are, respectively:

```sql
INSERT INTO t (a, b) VALUES (1, 2), (3, 4)
MERGE INTO t USING (...) ON (...) WHEN NOT MATCHED THEN INSERT (a, b) VALUES (...)
```

In native SQL, probably the least composable language ever, (I'm saying this without any COBOL experience), things look like they're the same when they're absolutely not. The two statements reuse a few keywords and syntax elements that make them look similar (`INTO`, `INSERT`, `VALUES`), but they're not at all the same.

Look at the language specification (the original docs by ISO are not free, but some free draft docs are available, see https://stackoverflow.com/q/6000847/521799). We have (only relevant bits):

### `INSERT`

```
<insert statement> ::= INSERT INTO <insertion target> <insert columns and source>

<insert columns and source> ::=
<from subquery>
| <from constructor>
| <from default>

<from constructor> ::=
[ <left paren> <insert column list> <right paren> ]
<contextually typed table value constructor>

<contextually typed table value constructor> ::=
VALUES <contextually typed row value expression list>

<contextually typed row value expression list> ::=
<contextually typed row value expression>
[ { <comma> <contextually typed row value expression> }... ]
```

### `MERGE`

```
<merge statement> ::=
MERGE INTO <target table> [ [ AS ] <merge correlation name> ]
USING <table reference>
ON <search condition> <merge operation specification>

<merge operation specification> ::=
<merge when clause>...

<merge when clause> ::=
<merge when matched clause>
| <merge when not matched clause>

<merge when not matched clause> ::=
WHEN NOT MATCHED [ AND <search condition> ]
THEN <merge insert specification>

<merge insert specification> ::=
INSERT [ <left paren> <insert column list> <right paren> ]
VALUES <merge insert value list>

<merge insert value list> ::=
<left paren>
<merge insert value element> [ { <comma> <merge insert value element> }... ]
<right paren>
```

As you can see, the two statements really have nothing to do with each other, syntactically, even if they share the common semantics of being able to `INSERT` into a table. With standard SQL `MERGE`, if you want to insert multiple rows, you have to put them in the `USING` clause, not in the `VALUES` clause.

Some RDBMS have attempted to implement an alternative way to `UPSERT` data (what you probably have in mind), compared to standard SQL `MERGE`. H2 is one of them, but support for the H2-specific `MERGE` statement, which looks much more like `INSERT` (http://h2database.com/html/commands.html#merge_into) will be deprecated in favour of the standard SQL one (http://h2database.com/html/commands.html#merge_using), starting from the next jOOQ minor release: https://github.com/jOOQ/jOOQ/issues/10045

It's a syntax that is supported only by H2 in this form (and in a somewhat similar form by HANA, via `UPSERT`) and has to be emulated everywhere else. Given these facts, and that H2 now supports standard SQL `MERGE USING`, the amount of jOOQ API purely dedicated to H2 `MERGE INTO` (46 interfaces!) should be removed.

**Comment by juanmirocks** [comment]
Dear @lukaseder THANK YOU so much for such a detailed answer. It's great to know the reasons why.

For my case and now, I guess I will just send multiple H2 `MERGE INTO` statements. It's simpler for me to reason about and I think the code also becomes simpler, in comparison to using the `MERGE USING` syntax.
